packages/muelu/src/Transfers/Matrix-Free/MueLu_MatrixFreeTentativeP_kokkos_def.hpp

```diff
@@ -0,0 +1,141 @@
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL SANDIA CORPORATION OR THE
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact
+//                    Jonathan Hu       (jhu@sandia.gov)
+//                    Andrey Prokopenko (aprokop@sandia.gov)
+//                    Ray Tuminaro      (rstumin@sandia.gov)
+//
+// ***********************************************************************
+//
+// @HEADER
+#ifndef MUELU_MATRIXFREETENTATIVEP_KOKKOS_DEF_HPP
+#define MUELU_MATRIXFREETENTATIVEP_KOKKOS_DEF_HPP
+
+#ifdef HAVE_MUELU_KOKKOS_REFACTOR
+
+#include "MueLu_MatrixFreeTentativeP_kokkos_decl.hpp"
+
+#include "MueLu_Aggregates_kokkos.hpp"
+#include "MueLu_AmalgamationFactory_kokkos.hpp"
+#include "MueLu_AmalgamationInfo_kokkos.hpp"
+#include "MueLu_CoarseMapFactory_kokkos.hpp"
+#include "MueLu_NullspaceFactory_kokkos.hpp"
+#include "MueLu_Level.hpp"
+#include "MueLu_MasterList.hpp"
+#include "MueLu_PerfUtils.hpp"
+#include "MueLu_PFactory.hpp"
+#include "MueLu_Monitor.hpp"
+#include "MueLu_Utilities_kokkos.hpp"
+
+#include <KokkosCompat_ClassicNodeAPI_Wrapper.hpp>
+
+#include "Teuchos_ScalarTraits.hpp"
+
+namespace MueLu {
+
+  // compute Y = alpha*R*X + beta*Y
+  template <class Scalar, class LocalOrdinal, class GlobalOrdinal, class DeviceType>
+  void MatrixFreeTentativeP_kokkos<Scalar,LocalOrdinal,GlobalOrdinal,Kokkos::Compat::KokkosDeviceWrapperNode<DeviceType>>::apply(const MultiVector &X, 
+                                          MultiVector &Y, 
+                                          Teuchos::ETransp mode, 
+                                          Scalar alpha, 
+                                          Scalar beta) const {
+
+    // Step 1: Y*=beta*Y, setup necessary structures
+    Y.scale(beta);
+
+    auto kokkos_view_X = X.getDeviceLocalView(Xpetra::Access::ReadOnly);
+    auto kokkos_view_Y = Y.getDeviceLocalView(Xpetra::Access::ReadWrite);
+    LO numCols = kokkos_view_X.extent(1);
+
+    if(mode == Teuchos::TRANS) { // if we're in restrictor mode
+      LO numAggs = aggregates_->GetNumAggregates();
+      Kokkos::View<LO*,DeviceType> aggPtr;
+      Kokkos::View<LO*,DeviceType> aggNodes;
+      Kokkos::View<LO*,DeviceType> unaggregated;
+      typename Aggregates_kokkos::aggregates_sizes_type::const_type aggSizes = aggregates_->ComputeAggregateSizes(); // TODO: can probably remove this and just use aggPtr if needed
+      aggregates_->ComputeNodesInAggregate(aggPtr,aggNodes,unaggregated);
```

Comment:
Couldn't you use `vertex2agg` here as well?

Reply:
I made a comment about `sqrt`ing the whole view on line 101 actually :stuck_out_tongue: It's probably slightly more performant.

As for `vertex2agg`, I need to think about this. For some reason I'm confused. I could have sworn I needed `GetVertex2AggId` for one direction and `ComputeNodesInAggregate` for the other direction, and now I'm questioning everything

Reply:
I decided against doing a sqrt of the aggSize vector for now. I don't think it would be easy to justify an entire kernel launch for sqrting the aggSizes